### PR TITLE
Add MetadataUpdateHandlerAttribute

### LINKED ISSUE #49361: Add a notification mechanism for EnC and HotReload
## Background and Motivation

One of the key work stream for the success of HotReload is a feedback mechanism so frameworks and 3rd party libraries knows that a type was hot reloaded and can decide to replace a live instance, invalidate a cache, refresh a binding, etc...

The proposed API isn't prescriptive, and other feedback mechanisms can be discussed, but it's quite important that the API is part of the BCL and the feature is supported by the runtime, so all frameworks will benefits from it, and it'll work in all HotReload scenarios (debugger and botnet watch).

This proposal was first mentioned in #45689, 2nd bullet of the 'Related API changes' section

## Proposed API

```C#
namespace System.Reflection.Metadata
{
    [System.AttributeUsage(System.AttributeTargets.Assembly, AllowMultiple = true)]
    public sealed class  ApplyUpdateNotificationAttribute : Attribute
    {
        /// <summary>
        /// HotReload notification API
        ///
        /// Declares which static method to execute when a HotReload delta is applied. The hot reload agent is responsible
        /// for scanning assemblies at load, and register the notifiers.
        ///
        /// notifier signature should look like: internal static void BeforeUpdate(Type type) and/or AfterUpdate(Type type).
        /// </summary>
        /// <param name="type">The Type on which the notifier is defined</param>
        /// <exception cref="ArgumentNullException">type or notifier are null</exception>
        /// <exception cref="...">type or notifier does not exist, or notifier isn't static, or doesn't have the right signature</exception>
        public ApplyUpdateNotificationAttribute (Type type)   
    }
}
```

## Usage Examples

This is how a UI Framework could reload a view when an HotReload change is applied

```C#
[assembly:ApplyUpdateNotification(typeof(HotReloadExtensions)]

namespace MyUIFramework
{
    public class View {
        public View Parent {get; set;} // setting the Parent sets the Child on the Parent view

        public View ()
        {
            HotReloadExtensions.RegisterInstance(this);
        }
    }

    internal static class HotReloadExtensions
    {
        // Keep weakRef of created views in a cache
        public static void RegisterInstance(View view) {...}

        // retrieve values from the cache
        static IEnumerable<View> GetViewsOfType(Type type) {...}

        // the notifier, declared by the attribute
        public static void AfterUpdate(Type type)
        {
            foreach (var oldinstance in GetViewsOfType(type)
            {
            }
        }
    }
}
```

## Alternative Designs

1. event: creating an event has the problem of registration, mainly for 3rd party libraries
2. implement this capability at the framework level. repetitive work, and the runtime would still have to implement it's own mechanism to invalide its caches (I'm thinking ReflectionCache, ....) 
3. whatever applies the change could also trigger the notification, but that would require injecting some shared code in all platforms supporting EnC/HR

## Risks

Unknown at this time

### PR DESCRIPTION
**Pull request by stephentoub**
And at least the beginning of reflection cache clearing support.

Fixes https://github.com/dotnet/runtime/issues/49361

cc: @tommcdon, @mikem8361, @steveharter, @pranavkm

### CODE REVIEW

**Review by mikem8361 — APPROVED** [review]
Other than a one comment, it looks good.

Looks like you also did the reflection caching clearing too.  Thank!

File: src/libraries/System.Runtime.Loader/tests/MetadataUpdateHandlerAttributeTest.cs [hunk 1]
```diff
@@ -0,0 +1,18 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Xunit;
+
```
[the remark this answers is not in the document]
Comment (stephentoub) [reply src/libraries/System.Runtime.Loader/tests/MetadataUpdateHandlerAttributeTest.cs:5]: Sorry, I don't understand the question... can you rephrase?

[the remark this answers is not in the document]
Comment (mikem8361) [reply src/libraries/System.Runtime.Loader/tests/MetadataUpdateHandlerAttributeTest.cs:5]: Never mind. I was confused about what you are testing here.

File: src/libraries/System.Private.CoreLib/src/System/Reflection/Metadata/MetadataUpdateHandlerAttribute.cs [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+namespace System.Reflection.Metadata
+{
+    /// <summary>Specifies a type that should receive notifications of metadata updates.</summary>
+    [AttributeUsage(AttributeTargets.Assembly, AllowMultiple = true)]
+    public sealed class MetadataUpdateHandlerAttribute : Attribute
+    {
+        /// <summary>Initializes the attribute.</summary>
+        /// <param name="handlerType">A type that handles metadata updates and that should be notified when any occur.</param>
+        public MetadataUpdateHandlerAttribute([DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.All)] Type handlerType) =>
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Reflection/Metadata/MetadataUpdateHandlerAttribute.cs:14]: @terrajobst, I deviated slightly from what's in the API issue here.  The argument was just named "type", and there wasn't a property to get the value.  I renamed it to "handlerType" and added a "HandlerType" property.  Any concerns?

Comment (pranavkm) [reply src/libraries/System.Private.CoreLib/src/System/Reflection/Metadata/MetadataUpdateHandlerAttribute.cs:14]: I imagine we would want to trim the handlers and associated hot reload code in a published / trimmed application. Does adding the DynamicallyAccessedMembers attribute make that more difficult?

File: src/libraries/System.Private.CoreLib/src/System/Reflection/Metadata/MetadataUpdateHandlerAttribute.cs [hunk 3]
```diff
@@ -0,0 +1,20 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics.CodeAnalysis;
+
+namespace System.Reflection.Metadata
+{
+    /// <summary>Specifies a type that should receive notifications of metadata updates.</summary>
```
Comment (pranavkm) [inline src/libraries/System.Private.CoreLib/src/System/Reflection/Metadata/MetadataUpdateHandlerAttribute.cs:8]: It would help to document the expected shape for the handler type here (if this is what ultimately ends up in the API docs).

Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Reflection/Metadata/MetadataUpdateHandlerAttribute.cs:8]: > It would help to document the expected shape for the handler type here 

It would.  We don't know it yet :)

File: src/libraries/System.Private.CoreLib/src/System/Reflection/Metadata/MetadataUpdateHandlerAttribute.cs (hunk 2 above)
Comment (eerhardt) [reply src/libraries/System.Private.CoreLib/src/System/Reflection/Metadata/MetadataUpdateHandlerAttribute.cs:14]: > As long as we tell the linker it should trim the MetadataUpdateHandlerAttribute when publishing

Who is going to do that? If not here, can we log an issue to track it?

> then that unroots things it might use dynamically

Correct, if the only thing holding onto `RuntimeTypeMetadataUpdateHandler` is the attribute referencing it `[assembly: MetadataUpdateHandler(typeof(RuntimeTypeMetadataUpdateHandler))]`, then if the trimmer was instructed to trim all `MetadataUpdateHandler` attributes, it can also remove the `RuntimeTypeMetadataUpdateHandler` class.

File: src/libraries/System.Runtime/tests/System/Reflection/ReflectionCacheTests.cs [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+using Xunit;
+
+namespace System.Reflection.Tests
+{
+    [SkipOnMono("Mono doesn't have a reflection cache")]
+    public class ReflectionCacheTests
+    {
+        [Fact]
+        public void GetMethod_MultipleCalls_SameObjects()
```
Comment (lambdageek) [inline src/libraries/System.Runtime/tests/System/Reflection/ReflectionCacheTests.cs:12]: nit: this test should pass on Mono.

File: src/libraries/System.Runtime.Loader/ref/System.Runtime.Loader.cs [hunk 5]
```diff
@@ -12,6 +12,12 @@ public static partial class AssemblyExtensions
         public unsafe static bool TryGetRawMetadata(this System.Reflection.Assembly assembly, out byte* blob, out int length) { throw null; }
         public static void ApplyUpdate(Assembly assembly, ReadOnlySpan<byte> metadataDelta, ReadOnlySpan<byte> ilDelta, ReadOnlySpan<byte> pdbDelta) { throw null; }
     }
+    [System.AttributeUsage(System.AttributeTargets.Assembly, AllowMultiple = true)]
+    public sealed class MetadataUpdateHandlerAttribute : System.Attribute
+    {
+        public MetadataUpdateHandlerAttribute([System.Diagnostics.CodeAnalysis.DynamicallyAccessedMembersAttribute(System.Diagnostics.CodeAnalysis.DynamicallyAccessedMemberTypes.All)] System.Type handlerType) { }
+        public System.Type HandlerType { get { throw null; } }
```
Comment (eerhardt) [inline src/libraries/System.Runtime.Loader/ref/System.Runtime.Loader.cs:19]: Typically if the ctor annotates the param as `DynamicallyAccessedMembers`, you will also want to annotate the property.

I assume there is going to be some caller that loops through all the `MetadataUpdateHandlerAttribute` instances, and starts calling methods on the `HandlerType` (using Reflection?). If that is the case, that caller is going to get ILLink warnings that methods on the `HandlerType` might be trimmed. So adding the attribute to the property will pre-emptively fix that.

File: src/libraries/System.Runtime/tests/System/Reflection/ReflectionCacheTests.cs [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+            Assert.NotNull(mi1);
+
+            MethodInfo mi2 = typeof(ReflectionCacheTests).GetMethod(nameof(GetMethod_MultipleCalls_SameObjects));
+            Assert.NotNull(mi2);
+
+            Assert.Same(mi1, mi2);
+        }
+
+        [Fact]
+        public void GetMethod_MultipleCalls_ClearCache_DifferentObjects()
```
Comment (lambdageek) [inline src/libraries/System.Runtime/tests/System/Reflection/ReflectionCacheTests.cs:24]: this test will fail in mono until https://github.com/dotnet/runtime/issues/50978 is done

File: src/libraries/System.Runtime.Loader/ref/System.Runtime.Loader.cs (hunk 5 above)
Comment (stephentoub) [reply src/libraries/System.Runtime.Loader/ref/System.Runtime.Loader.cs:19]: Ok.  I don't know if the caller is going to use the property or the ctor argument, and I don't expect the caller (the hot reload agent) to have trimming enabled, but it doesn't hurt to put the attribute in both places.

File: src/libraries/System.Runtime/tests/System/Reflection/ReflectionCacheTests.cs (hunk 4 above)
Comment (stephentoub) [reply src/libraries/System.Runtime/tests/System/Reflection/ReflectionCacheTests.cs:12]: Interesting.  I was under the impression mono didn't have a reflection cache.  Thanks for correcting me.

File: src/libraries/System.Private.CoreLib/src/System/Reflection/Metadata/MetadataUpdateHandlerAttribute.cs (hunk 2 above)
Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Reflection/Metadata/MetadataUpdateHandlerAttribute.cs:14]: Where/how do I say the attribute should be trimmed if Debugger.IsSupported is false?

Comment (eerhardt) [reply src/libraries/System.Private.CoreLib/src/System/Reflection/Metadata/MetadataUpdateHandlerAttribute.cs:14]: Yes, you should be able to add an entry in that file, just like the existing ones.

> is that just about trimming these out of corelib, or does that end up applying to all assemblies?

The way it works is the `<assembly fullname="System.Private.CoreLib"` identifies which assembly the attribute is defined in. These entries tell the trimmer to remove the attribute definition and all instantiations of the attribute in all assemblies. However, note that if the trimmer sees code reading the attribute, the trimmer will leave the attributes in, and won't trim them. The reasoning is that if there is code in the app reading the attribute, then it shouldn't be trimmed.

An issue with some attributes (like `NullableContextAttribute`) is that they can be defined in any assembly, the attribute isn't just defined in one assembly. In that case, we need to say `<assembly fullname="*"`. But here, this attribute should only ever be defined in CoreLib.

**Comment by stephentoub** [comment]
@pranavkm, I was just looking at dotnet-watch to see what we'll need to do there to respect this, and it seems like we likely first need Microsoft.CodeAnalysis.ExternalAccess.Watch.Api.WatchHotReloadService.Update augmented to include type names, then we can modify the manager to forward those to the agent, and then the agent can use those to resolve the Types to be passed to the update handlers?

File: src/libraries/System.Runtime.Loader/ref/System.Runtime.Loader.cs (hunk 5 above)
Comment (eerhardt) [inline src/libraries/System.Runtime.Loader/ref/System.Runtime.Loader.cs:19]: The implementation was updated to annotate the property, but the ref file wasn't.

```suggestion
        [System.Diagnostics.CodeAnalysis.DynamicallyAccessedMembersAttribute(System.Diagnostics.CodeAnalysis.DynamicallyAccessedMemberTypes.All)]
        public System.Type HandlerType { get { throw null; } }
```

**Review by eerhardt — APPROVED** [review]
LGTM - just one comment about updating the ref file.

File: src/libraries/System.Runtime.Loader/ref/System.Runtime.Loader.cs (hunk 5 above)
Comment (stephentoub) [reply src/libraries/System.Runtime.Loader/ref/System.Runtime.Loader.cs:19]: Oops. Thanks.

File: src/coreclr/System.Private.CoreLib/src/System/RuntimeType.CoreCLR.cs [hunk 7]
```diff
@@ -2418,6 +2418,37 @@ private RuntimeTypeCache InitializeCache()
...
+                // If the GCHandle doesn't wrap a cache yet, there's nothing more to do.
+                RuntimeTypeCache? existingCache = (RuntimeTypeCache?)GCHandle.InternalGet(m_cache);
+                if (existingCache is null)
+                {
+                    return;
+                }
+
+                // Create a new, empty cache to replace the old one and try to substitute it in.
+                var newCache = new RuntimeTypeCache(this);
+                if (ReferenceEquals(GCHandle.InternalCompareExchange(m_cache, newCache, existingCache), existingCache))
```
Comment (jkotas) [inline src/coreclr/System.Private.CoreLib/src/System/RuntimeType.CoreCLR.cs:2441]: It should be enough to just set the GCHandle target to null. The GCHandle is weakhandle and so the code should be generally prepared to handle the null case.

#51070
